src/main/src/main/java/org/geoserver/catalog/CatalogBuilder.java

```diff
@@ -1077,13 +1035,88 @@ private CoverageInfo buildCoverageInternal(GridCoverage2DReader reader, String n
 
-        /// dispose coverage 
-        gc.dispose(true);
-        if(gc.getRenderedImage() instanceof PlanarImage) {
-            ImageUtilities.disposePlanarImageChain((PlanarImage) gc.getRenderedImage());
+        return cinfo;
+    }
+
+    private GridSampleDimension[] getCoverageSampleDimensions(GridCoverage2DReader reader, Map customParameters)
+            throws TransformException, IOException, Exception {
+        GridEnvelope originalRange = reader.getOriginalGridRange();
+        Format format = reader.getFormat();
+        final ParameterValueGroup readParams = format.getReadParameters();
+        final Map parameters = CoverageUtils.getParametersKVP(readParams);
+        final int minX = originalRange.getLow(0);
+        final int minY = originalRange.getLow(1);
+        final int width = originalRange.getSpan(0);
+        final int height = originalRange.getSpan(1);
+        final int maxX = minX + (width <= 5 ? width : 5);
+        final int maxY = minY + (height <= 5 ? height : 5);
+
+        // we have to be sure that we are working against a valid grid range.
+        final GridEnvelope2D testRange = new GridEnvelope2D(minX, minY, maxX, maxY);
+
+        // build the corresponding envelope
+        final MathTransform gridToWorldCorner = reader.getOriginalGridToWorld(PixelInCell.CELL_CORNER);
+       
+        final GeneralEnvelope testEnvelope = CRS.transform(gridToWorldCorner, new GeneralEnvelope(testRange.getBounds()));
+        testEnvelope.setCoordinateReferenceSystem(reader.getCoordinateReferenceSystem());
+
+        if (customParameters != null) {
+        	parameters.putAll(customParameters);
         }
 
-        return cinfo;
+        // make sure mosaics with many superimposed tiles won't blow up with 
+        // a "too many open files" exception
+        String maxAllowedTiles = ImageMosaicFormat.MAX_ALLOWED_TILES.getName().toString();
+        if (parameters.keySet().contains(maxAllowedTiles)) {
+            parameters.put(maxAllowedTiles, 1);
+        }
+
+        // Since the read sample image won't be greater than 5x5 pixels and we are limiting the
+        // number of granules to 1, we may do direct read instead of using JAI
+        String useJaiImageRead = ImageMosaicFormat.USE_JAI_IMAGEREAD.getName().toString();
+        if (parameters.keySet().contains(useJaiImageRead)) {
+            parameters.put(useJaiImageRead, false);
+        }
+
+        parameters.put(AbstractGridFormat.READ_GRIDGEOMETRY2D.getName().toString(), new GridGeometry2D(testRange, testEnvelope));
+
+        // try to read this coverage
+        final GridCoverage2D gc = reader.read(CoverageUtils.getParameters(readParams, parameters, true));
+        final GridSampleDimension[] sampleDimensions;
+        if (gc != null) {
+            // remove read grid geometry since it is request specific
+            parameters.remove(AbstractGridFormat.READ_GRIDGEOMETRY2D.getName().toString());
```

Comment:
Indeed, makes no sense, parameters map is built in the method and does not get used outside.